From: Steve Dickson <SteveD@redhat.com>
Date: Tue, 2 Feb 2010 12:05:56 -0500
Subject: [nfs] nfsd41: nfsd4_decode_compound() does not recognize all ops
Message-id: <4B681524.1070209@RedHat.com>
Patchwork-id: 23088
O-Subject: [RHEL6] [PATCH 1/10] nfsd41: nfsd4_decode_compound() does not
	recognize all ops
Bugzilla: 560785
RH-Acked-by: Peter Staubach <staubach@redhat.com>
RH-Acked-by: Jeff Layton <jlayton@redhat.com>

commit c551866e649bac66a5145d100f34086d6edb581e
Author: Ricardo Labiaga <Ricardo.Labiaga@netapp.com>
Date:   Fri Dec 11 19:10:49 2009 -0800

    nfsd41: nfsd4_decode_compound() does not recognize all ops

    The server incorrectly assumes that the operations in the
    array start with value 0.  The first operation (OP_ACCESS)
    has a value of 3, causing the check in nfsd4_decode_compound
    to be off.

    Instead of comparing that the operation number is less than
    the number of elements in the array, the server should verify
    that it is less than the maximum valid operation number
    defined by LAST_NFS4_OP.

    Signed-off-by: Ricardo Labiaga <Ricardo.Labiaga@netapp.com>
    Signed-off-by: J. Bruce Fields <bfields@citi.umich.edu>

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/fs/nfsd/nfs4xdr.c b/fs/nfsd/nfs4xdr.c
index a8587e9..4f14f0c 100644
--- a/fs/nfsd/nfs4xdr.c
+++ b/fs/nfsd/nfs4xdr.c
@@ -1434,7 +1434,7 @@ nfsd4_decode_compound(struct nfsd4_compoundargs *argp)
 		}
 		op->opnum = ntohl(*argp->p++);
 
-		if (op->opnum >= OP_ACCESS && op->opnum < ops->nops)
+		if (op->opnum >= OP_ACCESS && op->opnum <= LAST_NFS4_OP)
 			op->status = ops->decoders[op->opnum](argp, &op->u);
 		else {
 			op->opnum = OP_ILLEGAL;
